drivers/gpio/gpio_nrfx.c

```diff
@@ -279,331 +218,159 @@ static int gpio_nrfx_port_toggle_bits(const struct device *port,
-					 uint32_t *sense_levels)
-{
-	struct gpio_nrfx_data *data = get_port_data(port);
-	const struct gpio_nrfx_cfg *cfg = get_port_cfg(port);
-	uint32_t level_pins = get_level_pins(port);
-	uint32_t port_in = nrf_gpio_port_in_read(cfg->port);
-
-	/* Extract which pins have logic level same as interrupt trigger level.
-	 */
-	uint32_t pin_states = ~(port_in ^ data->int_active_level);
-
-	/* Discard pins that aren't configured for level. */
-	uint32_t out = pin_states & level_pins;
-
-	/* Disable sense detection on all pins that use it, whether
-	 * they appear to have triggered or not.  This ensures
-	 * nobody's requesting DETECT.
-	 */
-	uint32_t pin = 0U;
-	uint32_t bit = 1U << pin;
-
-	uint32_t port_latch = 0;
-
-	uint32_t check_pins = level_pins;
-
-	if (IS_ENABLED(CONFIG_GPIO_NRF_INT_EDGE_USING_SENSE)) {
-		check_pins = data->pin_int_en;
+		trigger_config.p_in_channel = &ch;
 	}
 
-#if IS_ENABLED(CONFIG_GPIO_NRF_INT_EDGE_USING_SENSE)
-	/* Read LATCH, which will tell us which pin has changed its state. */
-	port_latch = cfg->port->LATCH;
-#endif
-
-	while (check_pins) {
-		if (check_pins & bit) {
-			uint32_t abs_pin = NRF_GPIO_PIN_MAP(cfg->port_num, pin);
-
-			if (!(level_pins & bit)) {
-				uint32_t sense = nrf_gpio_pin_sense_get(abs_pin);
-				bool high = (sense == GPIO_PIN_CNF_SENSE_High);
-
-				if (port_latch & bit) {
-					/* check if there was an interrupt */
-					if ((data->double_edge & bit) ||
-					    ((!!data->int_active_level) == high)) {
-						out |= bit;
-					}
-
-					/* invert configured level */
-					high = !high;
-				}
-
-				if (high) {
-					*sense_levels |= bit;
-				}
-			}
-
-			nrf_gpio_cfg_sense_set(abs_pin, NRF_GPIO_PIN_NOSENSE);
-			check_pins &= ~bit;
-		}
-		++pin;
-		bit <<= 1;
+	err = nrfx_gpiote_input_configure(abs_pin, NULL, &trigger_config, NULL);
+	if (err != NRFX_SUCCESS) {
+		return -EIO;
 	}
 
-#if IS_ENABLED(CONFIG_GPIO_NRF_INT_EDGE_USING_SENSE)
-	/* Clear LATCH, as at this point every level detection pin is
-	 * disabled.
-	 */
-	cfg->port->LATCH = port_latch;
-#endif
+	nrfx_gpiote_trigger_enable(abs_pin, true);
 
-	return out;
+	return 0;
 }
 
-static inline void fire_callbacks(const struct device *port, uint32_t pins)
+static int gpio_nrfx_manage_callback(const struct device *port,
+				     struct gpio_callback *callback,
+				     bool set)
 {
-	struct gpio_nrfx_data *data = get_port_data(port);
-	sys_slist_t *list = &data->callbacks;
-
-	gpio_fire_callbacks(list, port, pins);
+	return gpio_manage_callback(&get_port_data(port)->callbacks,
+				     callback, set);
 }
 
-static void gpiote_event_handler(void)
+/* Get port device from absolute pin number. */
+static const struct device *get_dev(uint32_t port_id)
 {
-	uint32_t fired_triggers[GPIO_COUNT] = {0};
-	uint32_t sense_levels[GPIO_COUNT] = {0};
-	bool port_event = nrf_gpiote_event_check(NRF_GPIOTE,
-						 NRF_GPIOTE_EVENT_PORT);
-
-	if (port_event) {
-#ifdef CONFIG_GPIO_NRF_P0
```

Comment:
I've just realized that so far it was possible to disable support for P0 or P1 in Kconfig. I'm not sure whether this possibility was actually used, but I guess it should be kept, so `get_dev()` should be reworked a bit.

Reply:
rebased on #38995 and updated `get_dev()` to use device tree.